File: backend/src/main/java/com/team01/project/domain/follow/service/QueryFollowService.java
package com.team01.project.domain.follow.service;

import java.util.List;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.team01.project.domain.follow.controller.dto.CountFollowResponse;
import com.team01.project.domain.follow.controller.dto.FollowResponse;
import com.team01.project.domain.follow.repository.FollowRepository;
import com.team01.project.domain.user.entity.User;
import com.team01.project.domain.user.repository.UserRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class QueryFollowService {

	private final FollowRepository followRepository;
	private final UserRepository userRepository;

	public List<FollowResponse> findFollowing(String currentUserId, String userId) {
		User currentUser = userRepository.getById(currentUserId);
		User user = userRepository.getById(userId);

		return followRepository.findByFromUser(user).stream()
			.map(follow -> FollowResponse.of(
				follow.getToUser(),
				checkFollow(follow.getToUser(), currentUser),
				checkFollow(currentUser, follow.getToUser())
			))
			.toList();
	}

	public List<FollowResponse> findFollower(String currentUserId, String userId) {
		User currentUser = userRepository.getById(currentUserId);
		User user = userRepository.getById(userId);

		return followRepository.findByToUser(user).stream()
			.map(follow -> FollowResponse.of(
				follow.getFromUser(),
				checkFollow(follow.getFromUser(), currentUser),
				checkFollow(currentUser, follow.getFromUser())
			))
			.toList();
	}

	public CountFollowResponse findCount(String userId) {
		User user = userRepository.getById(userId);
		Long followingCount = followRepository.countByFromUser(user);
		Long followerCount = followRepository.countByToUser(user);

		return CountFollowResponse.of(followingCount, followerCount);
	}



	private boolean checkFollow(User user, User currentUser) {
		return followRepository.existsByToUserAndFromUser(user, currentUser);
	}

	public Boolean checkMutualFollow(String currentUserId, String userId) {
		User user = userRepository.getById(userId);
		User currentUser = userRepository.getById(currentUserId);
		boolean checkFollower = followRepository.existsByToUserAndFromUser(currentUser, user);
		boolean checkFollowing = followRepository.existsByToUserAndFromUser(user, currentUser);

		return checkFollower && checkFollowing;
	}
}

File: backend/src/main/java/com/team01/project/domain/follow/service/QueryFollowService.kt
package com.team01.project.domain.follow.service

import com.team01.project.domain.follow.controller.dto.CountFollowResponse
import com.team01.project.domain.follow.controller.dto.FollowResponse
import com.team01.project.domain.follow.entity.Follow
import com.team01.project.domain.follow.entity.type.Status
import com.team01.project.domain.follow.repository.FollowRepository
import com.team01.project.domain.user.dto.SimpleUserResponse
import com.team01.project.domain.user.entity.User
import com.team01.project.domain.user.repository.UserRepository
import com.team01.project.domain.user.repository.findByIdOrThrow
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
class QueryFollowService(
    private val followRepository: FollowRepository,
    private val userRepository: UserRepository
) {

    fun findFollowing(currentUserId: String, userId: String): List<FollowResponse> {
        val currentUser = userRepository.findByIdOrThrow(currentUserId)
        val user = userRepository.findByIdOrThrow(userId)

        return followRepository.findByFromUserAndStatus(user, Status.ACCEPT).stream()
            .filter { follow: Follow -> follow.toUser.id != currentUser.id }
            .map { follow: Follow ->
                FollowResponse(
                    SimpleUserResponse.from(follow.toUser),
                    checkFollow(follow.toUser, currentUser),
                    checkFollow(currentUser, follow.toUser)
                )
            }
            .toList()
    }

    fun findFollower(currentUserId: String, userId: String): List<FollowResponse> {
        val currentUser = userRepository.findByIdOrThrow(currentUserId)
        val user = userRepository.findByIdOrThrow(userId)

        return followRepository.findByToUserAndStatus(user, Status.ACCEPT).stream()
            .filter { follow: Follow -> follow.fromUser.id != currentUser.id }
            .map { follow: Follow ->
                FollowResponse(
                    SimpleUserResponse.from(follow.fromUser),
                    checkFollow(follow.fromUser, currentUser),
                    checkFollow(currentUser, follow.fromUser)
                )
            }
            .toList()
    }

    fun findCount(userId: String): CountFollowResponse {
        val user = userRepository.findByIdOrThrow(userId)
        val followingCount = followRepository.countByFromUserAndStatus(user, Status.ACCEPT)
        val followerCount = followRepository.countByToUserAndStatus(user, Status.ACCEPT)

        return CountFollowResponse(followingCount, followerCount)
    }

    private fun checkFollow(user: User, currentUser: User): Status {
        return followRepository.findStatusByToUserAndFromUser(user, currentUser)
            .orElse(Status.NONE)
    }

    fun findMyFollowing(currentUserId: String): List<FollowResponse> {
        val currentUser = userRepository.findByIdOrThrow(currentUserId)

        return followRepository.findByFromUser(currentUser)
            .stream()
            .map { follow: Follow ->
                FollowResponse(
                    SimpleUserResponse.from(follow.toUser),
                    follow.status,
                    Status.NONE
                )
            }
            .toList()
    }

    fun findPendingList(currentUserId: String): List<FollowResponse> {
        val currentUser = userRepository.findByIdOrThrow(currentUserId)

        return followRepository.findByToUserAndStatus(currentUser, Status.PENDING).stream()
            .map { follow: Follow ->
                FollowResponse(
                    SimpleUserResponse.from(follow.fromUser),
                    Status.NONE,
                    follow.status
                )
            }
            .toList()
    }
}
